src/werkzeug/sansio/multipart.py

```diff
@@ -0,0 +1,251 @@
+                )
+                name = cast(str, extra.get("name"))
+                filename = extra.get("filename")
+                if filename is not None:
+                    event = File(
+                        filename=self._fix_ie_filename(filename),
+                        headers=headers,
+                        name=name,
+                    )
+                else:
+                    event = Field(
+                        headers=headers,
+                        name=name,
+                    )
+                self.state = State.DATA
+
+        elif self.state == State.DATA:
+            # Return up to the last line break as data, anything past
+            # that line break could be a boundary - more data may be
+            # required to know for sure.
+            lines = list(LINE_BREAK_RE.finditer(self.buffer))
+            if len(lines):
+                data_length = del_index = lines[-1].start()
+                match = self.boundary_re.search(self.buffer)
+                if match is not None:
+                    if match.group(1).startswith(b"--"):
+                        self.state = State.EPILOGUE
+                    else:
+                        self.state = State.PART
+                    data_length = match.start()
+                    del_index = match.end()
+
+                data = bytes(self.buffer[:data_length])
+                del self.buffer[:del_index]
+                more_data = match is None
+                if data or not more_data:
+                    event = Data(data=data, more_data=more_data)
+
+        elif self.state == State.EPILOGUE and self.complete:
+            event = Epilogue(data=bytes(self.buffer))
+            del self.buffer[:]
+            self.state = State.COMPLETE
+
+        if self.complete and isinstance(event, NeedData):
+            raise ValueError("Invalid form-data")
+
+        return event
+
+    def _parse_headers(self, data: bytes) -> Headers:
+        headers: List[Tuple[str, str]] = []
+        # Merge the continued headers into one line
+        data = HEADER_CONTINUATION_RE.sub(b" ", data)
+        # Now there is one header per line
+        for line in data.splitlines():
+            if line.strip() != b"":
+                name, value = _to_str(line).strip().split(":", 1)
+                headers.append((name.strip(), value.strip()))
+        return Headers(headers)
+
+    def _fix_ie_filename(self, filename: str) -> str:
+        """Internet Explorer 6 transmits the full file name if a file is
+        uploaded.  This function strips the full path if it thinks the
+        filename is Windows-like absolute.
+        """
+        if filename[1:3] == ":\\" or filename[:2] == "\\\\":
+            return filename.split("\\")[-1]
+
+        return filename
+
+
+class MultipartEncoder:
+    def __init__(self, boundary: bytes) -> None:
+        self.boundary = boundary
+        self.state = State.PREAMBLE
+
+    def encode(self, event: Event) -> bytes:
```

Comment:
I'd expect a method named `encode` to take a data structure to encode, not individual events. Perhaps this method should be private, and the rest of the implementation from `test.stream_encode_multipart` should be moved here as the `encode` method?

Reply:
I've added a to a discussion later about the API. The issue with trying to use the full implementation is that the File objects do IO, as the data is often stored in a file rather than in memory.

Reply:
Renamed to send_event, hopefully is now clearer and mirrors receive_data.